### ISSUE TITLE
EDU-1 — `tutorials/` scaffold + pure-function JAX MACE forward (energies)

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1607]
**Depends on:** — · **Blocks:** EDU-2 (#1608)

**Context:** MACE has no clean, readable baseline a computational chemist, a scientific-computing professional, a student, or an LLM can read to understand the model's mathematics and physics. The production code is not that artifact: the energy forward is spread across `mace/modules/models.py` (1459 lines), `blocks.py` (2061), `symmetric_contraction.py` (273) and `radial.py` (384), threaded through e3nn irreps machinery, backend dispatch (`wrapper_ops.py`), LAMMPS ghost-atom branches and `torch_geometric` batching — optimised for speed and for deployment, not for reading. `ScaleShiftMACE.forward` alone (`models.py:456-622`) carries the strain-displacement setup, the ML-IAP slicing, the joint-embedding branch and the pair-repulsion branch before it reaches a single message-passing layer. This ticket creates a separate, top-level `tutorials/` directory holding a single-file, pure-function **JAX** implementation of the MACE forward (energies only), written nanoGPT-style: readable over fast, the model passed in explicitly rather than hidden in closures. The target reader is assumed to have little-to-no ML-framework experience.

This is an **independent track with no phase gate** and **no dependency on the new v1 stack** — it can start immediately and does not import from `packages/`. It is deliberately *not* two things it could be confused with:

- **Not the "reference backend."** The reference backend is the plain-torch/plain-jax kernel oracle that lives *inside* the packages (`mace_torch/backends/reference/`, `mace_jax/kernels/reference.py`) and is a correctness oracle for real training/inference. `tutorials/` is a standalone teaching artifact, outside the supported packages, that nothing in `packages/` depends on.
- **Not a training path and not a mirror of the latest architecture.** It is **evaluation/inspection-only**: it exists so a human can read the model blocks and run a forward on a fixed set of parameters. There is **no JAX training anywhere** in the project — the `mace_jax` package itself is inference-only, and these tutorials are inspection/eval-only. Do not add an optimiser, a loss, or a training loop.

**Interface & constraints:**

- **Location & naming.** A top-level `tutorials/` directory, a **sibling of `packages/`** and the frozen legacy `mace/` — a repo-root directory, not a Python package. The name is decided: **`tutorials/`** (never `tiny_mace`, rejected specifically so users do not mistake it for a compressed/quantized production model). This ticket owns:
  ```
  tutorials/
  ├── mace_forward.py     # pure-function JAX MACE forward (energies), nanoGPT-style, no closures — EDU-1 (#1607)
  ├── notebooks/          # (created here as an empty scaffold; populated by EDU-2 (#1608))
  └── tests/              # (this ticket seeds a smoke test; EDU-2 (#1608) adds the oracle cross-check)
  ```

- **Language & style.** JAX. **Readability over performance**: prefer standard batch iteration (Python loops over atoms/edges/layers) to nested `vmap` where `vmap` would obfuscate the logic; omit performance-driven tricks (e.g. the fast spherical-harmonics recurrences) in favour of explicit closed-form code. **Descriptive variable names** — short math-style abbreviations are discouraged (write `spherical_harmonics`, `edge_lengths`, not `sh`, `r`). Inspirations to match in spirit: nanoGPT / llm.c, Psi4NumPy, tiny DFT / min_dft.

- **Pure functions, no closures (hard requirement).** The model is passed **explicitly** as data — a plain parameter/hyperparameter structure argument — into free functions. Do **not** adopt the JAX-MD-style API where an `init`/`step` closure pair hides the model state; that pattern is the explicit anti-example. Every block is a top-level function of the form `block(params, inputs) -> outputs`; the top-level energy function threads them together in the open.

- **The forward reproduces the energy `mace_run_train` models actually compute**, block-by-block, each a readable function so a reader can map code to the paper. The block list and its ordering are the ones on `develop`, not a stylised version of them:

  1. **Isolated-atom energies (E0).** `AtomicEnergiesBlock` (`blocks.py:363-390`) is a `[n_elements, n_heads]` table contracted with the one-hot element attributes; the per-node result is scatter-summed into a per-structure `e0`. This is a **term of the total energy**, not preprocessing, and omitting it makes every number wrong.
  2. **Node embedding** — `LinearNodeEmbeddingBlock` (`blocks.py:44-60`): one-hot element attributes → scalar (`0e`) node features.
  3. **Radial edge embedding** — `RadialEmbeddingBlock` (`blocks.py:394-435`): `BesselBasis` (`radial.py:18-57`, `sqrt(2/r_max) · sin(n π r / r_max) / r`, `n = 1…num_bessel`, a non-trainable buffer, default `--num_radial_basis 8`) multiplied by `PolynomialCutoff` (`radial.py:113-146`, the degree-`p` envelope with `--num_cutoff_basis` defaulting to **5**, hard-zeroed by `(x < r_max)`). With the default `apply_cutoff=True` the block returns `radial * cutoff` and a `None` cutoff; with `apply_cutoff=False` it returns the two separately and the interaction applies the cutoff itself. The tutorial implements the default branch and says so.
  4. **Angular edge embedding** — the **spherical harmonics** of the edge directions. Develop uses e3nn's `o3.SphericalHarmonics(sh_irreps, normalize=True, normalization="component")` (`models.py:163-165`) over `o3.Irreps.spherical_harmonics(max_ell)`, i.e. `0e+1o+2e+3o` at the default `--max_ell 3`. The tutorial writes them closed-form, but in **that** convention — real harmonics, component normalisation, unit-normalised directions — because a different normalisation is a silent constant per `ell` that EDU-2 (#1608)'s cross-check will fail on.
  5. **Message passing** — `--num_interactions` layers (default **2**). Each layer is an **interaction** (a channelwise tensor-product convolution over the neighbour graph, weighted by the radial MLP; `--interaction` and `--interaction_first` both default to `RealAgnosticResidualInteractionBlock`) followed by the **`EquivariantProductBasisBlock`** (`blocks.py:439-514`), whose `SymmetricContraction` (`symmetric_contraction.py:26-87`) is the many-body term up to `--correlation` (default **3**). Written as explicit sums over neighbours, not fused kernels.
  6. **Readout** — per-layer site energies. The last layer uses `readout_cls` (default `NonLinearReadoutBlock`), earlier layers use `LinearReadoutBlock`, and every readout emits one column per head, indexed by `node_heads`.
  7. **Reduction, in develop's order.** Sum the per-node readout terms; apply `ScaleShiftBlock` (`blocks.py:1835-1850`, per-head `scale[head] * x + shift[head]`) **to the per-node interaction energy** (`models.py:577-578`); scatter-sum over the atoms of each structure (`models.py:579`); then add `e0` **outside** the scaling (`models.py:581`). The scale/shift is not applied to the structure total and E0 is not scaled.

  The output is per-structure total energy (a scalar per input structure). Forces/stress are out of scope (they would need the autograd/strain glue and are not needed to *read* the model).

- **Implement `ScaleShiftMACE`, not `MACE`.** The plain `MACE` class exists in `models.py:47-440` but nothing in `mace/` instantiates it: `--model MACE`, the CLI default, builds a **`ScaleShiftMACE`** with `atomic_inter_shift=[0.0] * len(heads)` (and coerces `interaction_first` to `RealAgnosticInteractionBlock` unless it is one of three allowed values), while `--model ScaleShiftMACE` builds the same class with the dataset mean as shift (`mace/tools/model_script_utils.py:279-322`). The plain class also *orders the energy differently* — it scatter-sums each readout separately and stacks `e0`, the pair term and the readouts as unscaled contributions (`models.py:397-400`). A tutorial that implements it would be teaching a path no trained model takes.

- **Scope table.** The tutorial covers one path through develop's model surface. Everything else is dropped **for the tutorial only**; every row below is KEEP for v1 and is owned by the ticket named in its reason, so a reader must not read this table as a v1 DROP list.

| Develop capability | `tutorials/mace_forward.py` | Disposition |
|---|---|---|
| E0 term, node embedding, Bessel×cutoff, spherical harmonics, interaction + symmetric contraction, per-layer readout, per-head scale/shift | implemented | COVER |
| Forces, virials, stress, hessians, edge forces, atomic stresses (`get_outputs`, `mace/modules/utils.py:270`) | absent | DROP — needs the autograd/strain glue; reading the model does not (ARCH-4 (#1563) owns them for v1) |
| ZBL pair repulsion (`--pair_repulsion`, `radial.py:149-223`) | absent | DROP — off by default; an additive term that teaches nothing about equivariance |
| Multi-head readouts and head selection | single head only | DROP — the head index is a column selector, not a block (FT-1 (#1580)) |
| `AtomicDipolesMACE`, `AtomicDielectricMACE`, `EnergyDipolesMACE` | absent | DROP — same skeleton, different readout (RET-4 (#1600) / ELEC-2 (#1592)) |
| Magnetic family: `MagneticScaleShiftMACE`, `MagneticSCFMACE`, magmom as an input feature, magforces as an output, the SCF fixed point (`mace/modules/extensions.py:1428-2105`) | absent | DROP — a second equivariant channel and a fixed-point solver; out of a first reading (MAG-1 (#1584)) |
| Electrostatics: `MACELES` (`mace/modules/extensions.py:142-660` — BEC, `latent_charges`/`latent_dipoles`/`latent_alphas`/`latent_kappas`/`latent_quads`, `external_field`) with its calculator-side `eps_infty` / `keep_neutral` handling (`mace/calculators/mace.py:816-845`), and `PolarMACE` k-space | absent | DROP — long-range solvers are their own subject (ELEC-1 (#1591) / ELEC-2 (#1592)) |
| Joint embedding / `embedding_specs` / `use_embedding_readout` graph-level inputs | absent | DROP — an additive embedding term, not a block of the equivariant core (ARCH-3 (#1562)) |
| LoRA (`mace/modules/lora.py`), distance transforms (`AgnesiTransform`, `SoftTransform`), `use_so3`, `use_agnostic_product`, `edge_irreps` | absent | DROP — hyperparameter variants of blocks the tutorial already shows |
| cueq / oeq backend dispatch, `torch.compile`, TorchScript, LAMMPS ghost-atom branches | absent | DROP — the tutorial has no backends and no deployment target |

- **Independence invariants.** `tutorials/mace_forward.py` imports only JAX/NumPy (and stdlib); it imports nothing from `packages/`, nothing from the legacy `mace/`, and neither `torch` nor `e3nn`. It is not on any import path of `packages/`. Parameters are supplied to the tutorial as plain arrays via a small fixed fixture; producing/validating those parameters against a real MACE model is EDU-2 (#1608)'s cross-check, not this ticket.

**Task:**
1. Create the top-level `tutorials/` directory (sibling of `packages/`), with `notebooks/` and `tests/` subdirectories seeded (empty/placeholder for EDU-2 (#1608)).
2. Write `tutorials/mace_forward.py`: the pure-function JAX MACE energy forward, one readable function per block in the order above (E0 table, node embedding, Bessel×cutoff, closed-form component-normalised spherical harmonics, interaction/channelwise TP conv, symmetric contraction, per-layer readout, per-node scale/shift → scatter-sum → `+ e0`), model passed explicitly, no closures, descriptive names, standard loops over `vmap` where clarity wins.
3. Add a small self-contained fixture (a tiny molecule + a compatible parameter set) so the forward runs end-to-end without any external artifact or the new stack. The fixture records `r_max`, `num_bessel`, `num_polynomial_cutoff`, `max_ell`, `num_interactions`, `correlation`, and which Clebsch-Gordan basis its symmetric-contraction weights were built against — `SymmetricContraction` takes `use_reduced_cg`, `MACE.__init__` defaults it to `True`, and the CLI flag `--use_reduced_cg` defaults to `False` (`mace/tools/arg_parser.py:239-244`), so a CLI-trained model carries the **full** basis and the two are not interchangeable.
4. Seed `tutorials/tests/` with a smoke test that runs the forward on the fixture and checks the output shape/finiteness (the numerics cross-check against a real MACE is EDU-2 (#1608)).
5. Add a short `tutorials/README` stating what this is (educational, eval/inspection-only, JAX) and — explicitly — what it is **not** (not the reference backend, not a training path, not a mirror of the latest architecture, no JAX training), plus the scope table above so a reader knows what the file deliberately omits.

**Out of scope:** the marimo notebooks (spherical-harmonics / Clebsch-Gordan / message-passing / many-body math walkthrough), the CI job that executes scripts + notebooks, and the numerics cross-check against a real MACE forward on a golden fixture — all EDU-2 (#1608). Any training loop, optimiser, loss, forces/stress, or JAX-MD-style closure API. Any dependency on `packages/` or the v1 stack. Every row marked DROP in the scope table.

**Acceptance criteria:**
- [ ] `tutorials/` exists at the repo root as a sibling of `packages/`, containing `mace_forward.py`, `notebooks/`, and `tests/`; the directory is named `tutorials/` (not `tiny_mace`).
- [ ] `tutorials/mace_forward.py` computes per-structure MACE **energies** in pure JAX, one readable function per block, with the model/parameters passed in explicitly — no `init`/`step` closures hiding state.
- [ ] The energy is assembled in develop's order: per-node readout sum → per-head scale/shift → scatter-sum per structure → **plus** the E0 term, which is outside the scale/shift.
- [ ] The spherical harmonics are real and component-normalised over unit directions, matching `o3.SphericalHarmonics(..., normalize=True, normalization="component")`.
- [ ] The forward runs end-to-end on the in-file fixture and returns finite energies of the expected shape; the fixture records the hyperparameters and the Clebsch-Gordan basis its weights assume.
- [ ] `tutorials/mace_forward.py` imports nothing from `packages/`, the legacy `mace/`, `torch` or `e3nn` (asserted); it uses only JAX/NumPy/stdlib.
- [ ] No optimiser, loss, or training loop anywhere in `tutorials/`; forces/stress are absent by design.
- [ ] The README states the artifact is educational and eval/inspection-only, disclaims being the reference backend, a training path, or an architecture mirror, and carries the scope table marking the omissions as tutorial-scope, not v1 drops.

**Verify:**
```bash
python tutorials/mace_forward.py                          # runs the fixture forward, prints energies
python -m pytest tutorials/tests -v                       # smoke: shape + finiteness
python - <<'PY'                                           # independence guard
import ast, pathlib
src = pathlib.Path("tutorials/mace_forward.py").read_text()
tree = ast.parse(src)
mods = [n.module or "" for n in ast.walk(tree) if isinstance(n, ast.ImportFrom)]
mods += [a.name for n in ast.walk(tree) if isinstance(n, ast.Import) for a in n.names]
forbidden = {"mace", "mace_core", "mace_torch", "mace_jax", "torch", "e3nn"}
assert not any(m.split(".")[0] in forbidden for m in mods), mods
print("independent: OK")
PY
```
**Review focus:** readability first — a chemist/physicist with no ML-framework background should be able to follow each block from code to equation; that the energy assembly matches develop's order (scale/shift on per-node interaction energies, E0 added outside it) rather than a tidier-looking one; that the API is genuinely pure-function (model passed explicitly, no closures) and the file is self-contained (no `packages/`/`mace/`/torch/e3nn imports, no hidden training path).

---
